optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py

```diff
@@ -0,0 +1,197 @@
+            for key, val in variant.items():
+                variant_tree_nodes.append(TreeNode(key, {key: val}))
+
+            yield {"variant_id": vid,
+                   "variant": variant_tree_nodes,
+                   "paths": self.parameter_path}
+
+    def __len__(self):
+        return sum(1 for _ in self)
+
+    def update_defaults(self, defaults):
+        pass
+
+    def to_str(self, summary, variants, **kwargs):
+        """
+        Return human readable representation
+
+        The summary/variants accepts verbosity where 0 means silent and
+        maximum is up to the plugin.
+
+        :param summary: How verbose summary to output (int)
+        :param variants: How verbose list of variants to output (int)
+        :param kwargs: Other free-form arguments
+        :rtype: str
+        """
+        if not self.variants:
+            return ""
+
+        out = []
+        verbose = variants > 1
+        out.append("CIT Variants (%i):" % len(self))
+        for variant in self:
+            out.append('%sVariant %s:    %s' % ('\n' if verbose else '',
+                                                variant["variant_id"],
+                                                self.parameter_path))
+            if not verbose:
+                continue
+            env = set()
+            for node in variant["variant"]:
+                for key, value in iteritems(node.environment):
+                    origin = node.environment.origin[key].path
+                    env.add(("%s:%s" % (origin, key), str(value)))
+            if not env:
+                return out
+            fmt = '    %%-%ds => %%s' % max([len(_[0]) for _ in env])
+            for record in sorted(env):
+                out.append(fmt % record)
+
+        return "\n".join(out)
+
+
+class Cit(object):
+    def __init__(self, parameters, constraints):
+        self.parameters = parameters
+        # Constraints are not implemented in this example
+        self.constraints = constraints
+
+    def combine(self):
+        """
+        Example of method to combine the parameters. Will be
+        replaced by the actual CIT algorithm.
+
+        Example:
+           p1: x1, x1
+           p2: y1, y2
+
+           return (['p1', 'p2'],
+                   [{'p1': 'x1', 'p2': 'y1'},
+                    {'p1': 'x1', 'p2': 'y2'},
+                    {'p1': 'x2', 'p2': 'y1'},
+                    {'p1': 'x2', 'p2': 'y2'}])
+
+        :return: headers (list of parameters keys) and combinations (list of
+                 dictionaries. Each dictionary represents a combination of
+                 parameters.
+        :rtype: tuple
+        """
+        combinations = []
+        headers = [item[0] for item in self.parameters]
+        for combination in itertools.product(*(params[1]
+                                               for params in self.parameters)):
+            combinations.append(dict(zip(headers, combination)))
+        return headers, combinations
```

Comment:
have you considered adding some example selftest?

Reply:
Not at this point. Also, the actual combination code is tricky to test, considering it generates different combinations in each execution. I'd leave selftests to be created with the actual code in place.